kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractSimultaneousConnectionsTest.java

```diff
@@ -156,41 +151,90 @@ public void http1WebSocketConnections() throws Exception {
+    final CountDownLatch latch = new CountDownLatch(MAX_HTTP_1_WS_CONNECTIONS);
     final MockResponse response = new MockResponse()
         .withWebSocketUpgrade(new WebSocketListener() {
           @Override
           public void onOpen(okhttp3.WebSocket webSocket, Response response) {
+            try {
+              cyclicBarrier.await(1, TimeUnit.SECONDS);
+            } catch (Exception ignore) {
+            }
             serverSockets.add(webSocket);
             webSocket.send("go on");
           }
         });
-    IntStream.range(0, MAX_HTTP_1_CONNECTIONS).forEach(i -> mockWebServer.enqueue(response));
+    IntStream.range(0, MAX_HTTP_1_WS_CONNECTIONS).forEach(i -> mockWebServer.enqueue(response));
     try (final HttpClient client = clientBuilder.build()) {
-      for (int it = 0; it < MAX_HTTP_1_CONNECTIONS; it++) {
+      for (int it = 0; it < MAX_HTTP_1_WS_CONNECTIONS; it++) {
         client.newWebSocketBuilder()
             .uri(mockWebServer.url("/").uri())
             .buildAsync(new WebSocket.Listener() {
 
               @Override
               public void onMessage(WebSocket webSocket, String text) {
+                clientSockets.add(webSocket);
                 latch.countDown();
                 webSocket.request();
               }
-            })
-            .whenComplete((ws, t) -> clientSockets.add(ws));
+            });
+        cyclicBarrier.await(1, TimeUnit.SECONDS);
       }
       assertThat(latch.await(20L, TimeUnit.SECONDS)).isTrue();
       assertThat(serverSockets.size())
-          .isEqualTo(MAX_HTTP_1_CONNECTIONS)
+          .isEqualTo(MAX_HTTP_1_WS_CONNECTIONS)
           .isLessThanOrEqualTo((int) serverSocketFactory.activeConnections());
+      assertThat(clientSockets)
+          .hasSize(MAX_HTTP_1_WS_CONNECTIONS);
     } finally {
       for (okhttp3.WebSocket socket : serverSockets) {
         socket.close(1000, "done");
       }
     }
   }
 
+  private static class DelayedResponseHandler implements HttpHandler {
+
+    private final int requestCount;
+    private final CyclicBarrier barrier;
+    private final Set<HttpExchange> exchanges;
+    private final AtomicInteger connectionCount;
+    private final CompletableFuture<Void> done;
+
+    private DelayedResponseHandler(int requestCount, CyclicBarrier barrier, HttpHandler handler) {
+      this.requestCount = requestCount;
+      this.barrier = barrier;
+      exchanges = ConcurrentHashMap.newKeySet();
+      connectionCount = new AtomicInteger(0);
+      done = new CompletableFuture<>();
+      done.thenRunAsync(() -> {
+        for (HttpExchange exchange : exchanges) {
+          try {
+            handler.handle(exchange);
+          } catch (IOException ignore) {
+            // NO OP
+          }
+        }
+      });
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+      connectionCount.incrementAndGet();
+      try {
+        barrier.await(1, TimeUnit.SECONDS);
```

Comment:
I'm fine with using a mechanism like this rather than retries - I was not sure how intentional the flooding was, so I preserved that initial behavior just in case that was something you were trying to test.

You can also consider using a Semaphore instead of CyclicBarrier as there's no need to wait here, just increase the permits for the loop.

Reply:
The await is just a safety mechanism to make sure that the tests end as soon as possible in case there are any other non-contemplated errors that may hang the its execution. This way the test will halt from any of the sides in case one of them becomes unresponsive.
Maybe a semaphore is better suited for this, but don't know exactly how to implement the same safety features.

Reply:
Switching to semaphore was not creating enough back-pressure for the JDK implementation which failed randomly (about 25% of times in my local system), especially for `http1WebSocketConnectionsBeforeUpgrade`.
I've switched back to Cyclic Barrier. The slight block in the handler seems to do the job (ran same test >500 times with 0 failures locally -even with RejectedExecutions which caused test failures with Semaphore-)

Reply:
So even with a semaphore the re-try logic from the other pr was still necessary.  Long story short - the HttpServer can't handle very much - hopefully this will be reliable enough for now in github runs, but it does make a good case for moving the retry logic into our standardhttpclient to be more agnostic to this situation.

Reply:
Yes, perfectly summarized.